src/client/storage/index.js

```diff
@@ -68,32 +77,56 @@ const getChunk = async (chunkId, encoding = 'utf8', useCache = true) => {
         log.debug({chunkId}, 'Graphql request success');
 
         for (const edge of queryResult.transactions.edges) {
-            const txid = edge.node.id;
-            log.debug({chunkId, txid}, 'Downloading data from arweave');
+            try {
+                const txid = edge.node.id;
+                log.debug({chunkId, txid}, 'Downloading data from arweave');
+
+                const data = await storage.getDataByTxId(txid);
+                log.debug({chunkId, txid}, 'Successfully downloaded data from arweave');
+
+                const buf = Buffer.from(data);
+
+                const hash = hashFn(buf).toString('hex');
+                if (hash !== chunk.id) {
+                    log.warn(
+                        {chunkId, hash, query, buf: buf.toString()},
+                        'Chunk id and data do not match'
+                    );
+                    continue;
+                }
+
+                await fs.writeFile(chunkPath, buf);
+
+                chunk.size = buf.length;
+                chunk.dl_status = CHUNK_DOWNLOAD_STATUS.COMPLETED;
+                await chunk.save();
+                return buf;
+            } catch (err) {
+                log.warn({chunkId, err}, 'Trying to download chunk from Arweave threw an error');
+            }
+        }
 
-            const data = await storage.getDataByTxId(txid);
-            log.debug({chunkId, txid}, 'Successfully downloaded data from arweave');
+        log.warn(
+            {chunkId},
+            'Chunk not found in Arweave. Falling back to requesting it from bundler backup'
+        );
 
+        // TODO: refactor before mainnet!!!
+        // If we failed to retrieve the chunk from Arweave,
+        // retrieve it from the bundler's backup (S3).
+        try {
+            const url = `${config.get('storage.arweave_bundler_url')}/download/${chunkId}`;
+            const {data} = await axios.get(url);
+            log.debug({chunkId}, 'Failed to find it in Arweave but found it in the bundler');
             const buf = Buffer.from(data);
-
-            const hash = hashFn(buf).toString('hex');
-            if (hash !== chunk.id) {
-                log.warn(
-                    {chunkId, hash, query, buf: buf.toString()},
-                    'Chunk id and data do not match'
-                );
-                continue;
-            }
-
             await fs.writeFile(chunkPath, buf);
-
             chunk.size = buf.length;
             chunk.dl_status = CHUNK_DOWNLOAD_STATUS.COMPLETED;
             await chunk.save();
             return buf;
+        } catch (err) {
+            throw new Error('No matching hash found in Arweave nor bundler');
         }
-
-        throw new Error('No matching hash found in arweave');
     } catch (e) {
         log.error({chunkId, message: e.message, stack: e.stack}, 'Chunk download failed');
```

Comment:
Here's where the fallback should be, I think

Reply:
I'll move it.